From 4feb5e232207b8a79f02908f879b1174c25a4f36 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Tue, 29 Oct 2024 23:17:28 +0800
Subject: [PATCH 193/317] AOSCOS: drm: loongson: add ls7a1000_support module
 parameter

The loongson DRM driver is known to cause boot failure or hang on certain
LS7A1000-based hardware. Add a switch (`loongson.ls7a1000_support',
disabled by default to help users of LS7A2000/2K2000-based hardware
leverage the loongson DRM driver.

For those who feel adventurous, use the `loongson.ls7a1000_support=1'
parameter to enable this driver.

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

Link: https://t.me/c/1109254909/605851
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/gpu/drm/loongson/loongson_module.c | 4 ++++
 drivers/gpu/drm/loongson/loongson_module.h | 1 +
 drivers/gpu/drm/loongson/lsdc_drv.c        | 8 ++++++++
 3 files changed, 13 insertions(+)

diff --git a/drivers/gpu/drm/loongson/loongson_module.c b/drivers/gpu/drm/loongson/loongson_module.c
index d2a51bd395f6..89915a33ec71 100644
--- a/drivers/gpu/drm/loongson/loongson_module.c
+++ b/drivers/gpu/drm/loongson/loongson_module.c
@@ -17,6 +17,10 @@ int loongson_vblank = 1;
 MODULE_PARM_DESC(vblank, "Disable/Enable hw vblank support");
 module_param_named(vblank, loongson_vblank, int, 0400);
 
+int loongson_ls7a1000_support = 0;
+MODULE_PARM_DESC(ls7a1000_support, "7A1000 support (1 = enabled, 0 = disabled (default))");
+module_param_named(ls7a1000_support, loongson_ls7a1000_support, int, 0400);
+
 static int __init loongson_module_init(void)
 {
 	if (!loongson_modeset || video_firmware_drivers_only())
diff --git a/drivers/gpu/drm/loongson/loongson_module.h b/drivers/gpu/drm/loongson/loongson_module.h
index 931c17521bf0..0dcc9646d2ff 100644
--- a/drivers/gpu/drm/loongson/loongson_module.h
+++ b/drivers/gpu/drm/loongson/loongson_module.h
@@ -7,6 +7,7 @@
 #define __LOONGSON_MODULE_H__
 
 extern int loongson_vblank;
+extern int loongson_ls7a1000_support;
 extern struct pci_driver lsdc_pci_driver;
 
 #endif
diff --git a/drivers/gpu/drm/loongson/lsdc_drv.c b/drivers/gpu/drm/loongson/lsdc_drv.c
index b350bdcf1645..04c70c6157f9 100644
--- a/drivers/gpu/drm/loongson/lsdc_drv.c
+++ b/drivers/gpu/drm/loongson/lsdc_drv.c
@@ -267,6 +267,14 @@ static int lsdc_pci_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
 	int ret;
 
 	descp = lsdc_device_probe(pdev, ent->driver_data);
+
+	if (to_loongson_gfx(descp)->chip_id == CHIP_LS7A1000 &&
+	    !loongson_ls7a1000_support) {
+		dev_info(&pdev->dev,
+			"LS7A1000 support disabled or not enabled via module parameter\n");
+		return -ENODEV;
+	}
+
 	if (IS_ERR_OR_NULL(descp))
 		return -ENODEV;
 
-- 
2.48.1

